rust-tdshim/src/mp.rs

```diff
@@ -157,55 +159,53 @@ fn parallel_accept_memory(apic_id: u64) {
     }
 }
 
+// Wait for APs to arrive by checking if they are available
 fn wait_for_ap_arrive(cpu_num: u32) {
     let mut mail_box = MailBox::new(get_mem_slice_mut(SliceType::MailBox));
-    log::info!("Waiting for APs to arrive...\n");
-    loop {
-        if mail_box.cpu_arrival() == cpu_num - 1 {
-            log::info!("All APs has arrived\n");
-            break;
-        }
-    }
-    mail_box.set_cpu_exiting(cpu_num);
-}
-
-fn wait_for_ap_exit(cpu_num: u32) {
-    let mail_box = MailBox::new(get_mem_slice_mut(SliceType::MailBox));
-    log::info!("Waiting for APs to exit...\n");
-    loop {
-        let cpu_exit = mail_box.cpu_exiting();
-
-        if cpu_exit == 1 {
-            log::info!("All APs has exited\n");
-            break;
+    for i in 1..cpu_num {
+        mail_box.set_apic_id(i);
+        mail_box.set_command(MP_WAKEUP_COMMAND_AVAILABLE);
+
+        loop {
+            if mail_box.apic_id() == MAILBOX_APICID_INVALID {
+                mail_box.set_command(MP_WAKEUP_COMMAND_NOOP);
+                break;
+            } else {
+                cpu_pause();
+            }
         }
     }
+    x86::fence::mfence();
 }
 
 pub fn ap_assign_work(apic_id: u32, stack: u64, entry: u32) {
     let mut mail_box = MailBox::new(get_mem_slice_mut(SliceType::MailBox));
     mail_box.set_wakeup_vector(entry);
     mail_box.set_fw_arg(0, stack);
 
-    mail_box.set_apid_id(apic_id);
+    mail_box.set_apic_id(apic_id);
     mail_box.set_command(MP_WAKEUP_COMMAND_ACCEPT_PAGES);
 
     loop {
-        let wakeup_apic_id = mail_box.apid_id();
-        if wakeup_apic_id == MAILBOX_APICID_INVALID {
+        if mail_box.apic_id() == MAILBOX_APICID_INVALID {
+            mail_box.set_command(MP_WAKEUP_COMMAND_NOOP);
             log::info!("Successfully wakeup AP #{}\n", apic_id);
             break;
+        } else {
+            cpu_pause();
```

Comment:
This code is duplicated in `wait_for_ap_arrive()`. It might be worth creating a new function for this loop that both functions call to remove the risk of future maintenance issues...?

Reply:
Agree, will create a new function to do the common work.